### ISSUE TITLE
[ENH] support fitting first levels to numpy arrays directly, in addition to niimgs

### ISSUE THREAD

**Issue by henrymj** [nilearn/nilearn#4110]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe your proposed enhancement in detail.

We recently fit GLMs to edge timeseries (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10369861/) in order to look at fluctuations of attention at the network level. We believe it's an exciting approach that offers novel information and which is really easy to implement.

Almost all of the code was based off of nilearn - it mostly involved just changing FirstLevelModel().fit to take numpy arrays, rather than niimgs, so that we could give it the 2D arrays (n_features x n_scans). 

It would be really great to bake this feature directly into nilearn, so that it can be easily kept up to date with other nilearn features and bug-fixes. I believe the changes to the code would be minimal, and add little-to-no maintenance, and I would be happy to submit the pull request myself.

I have some example pieces of potential changes below.

### Benefits to the change

This would add flexibility to the nilearn code for fitting first-level models - letting it fit edges, or even just data that is in a significantly different format (e.g., open data that was already preprocessed and parcellated with some atlas). It would cost very little to maintain. 

### Pseudocode for the new behavior, if applicable

```python
# insert your code below

def fit(...):

if (self.mask_img is None) & (type(run_imgs[0])==np.ndarray):
    assert run_imgs[0].shape==2  # nFeatures x nScans
...

if type(run_img)==np.ndarray:
    Y = run_img
else:
    run_img = check_niimg(run_img, ensure_ndim=4)
    Y = self.masker_.transform(run_img, sample_mask=sample_mask)

if design_matrices is None:
     n_scans = Y.shape[-1]
     
     
def compute_contrast(...):

...

if self.masker_ is None:
    output = estimate_
else:
    output = self.masker_.inverse_transform(estimate_)



```

**Comment by bthirion** [comment]
Given that the GLM is natively implemented on arrays, it is indeed easy to do. The only that has to be done is to create a high-level API that encapsulates the function. So the cost is minimal.
I don't think we should advertise much these functions, but I think it is OK to have them.
Any other opinion ?

**Comment by Remi-Gau** [comment]
> Any other opinion ?

I think that sounds like a sensible approach. 

@henrymj would you be willing to start a pull request to start implementing this?

**Comment by henrymj** [comment]
Absolutely!

To clarify: @bthirion are you suggesting making a new function specifically designed to take lists of numpy arrays? My initial idea was to just modify `FirstLevelModel.fit` and `.compute_contrast` to be able to handle numpy arrays, in addition to niimgs.

**Comment by bthirion** [comment]
I think it would increase a lot the complexity of the object.  It also makes debugging difficult, when you hit an error in a complex pipeline and are not sure whether a given argument is an image, an array or whatever.

**Comment by Remi-Gau** [comment]
Realising that we already have run_glm that runs on numpy arrays.

https://nilearn.github.io/stable/modules/generated/nilearn.glm.first_level.run_glm.html#nilearn-glm-first-level-run-glm

Which is what we use in one of our surface example: https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_localizer_surface_analysis.html#perform-first-level-analysis

So I am wondering if we should have more support for numpy arrays.

**Comment by bthirion** [comment]
Yes, at the end of the data, we only work with Numpy arrays, but the point is whether to expose it, which I don't wish to do.

**Comment by Remi-Gau** [comment]
> the point is whether to expose it

yeah I remember our discussion about this.

what I was trying to say is that we already have a function that is part of the public user facing API that runs on numpy arrays, so my opinion is we should close this PR and its associated issue as there is already a public API function to do it.
